docker/Dockerfile

```diff
@@ -13,47 +13,129 @@
+RUN apt-get update \
+    && apt-get install -y software-properties-common \
+    && apt-get update && apt-get install -y --no-install-recommends \
+    autoconf \
+    bison \
+    build-essential \
+    ca-certificates \
+    ccache \
+    clang-format-7 \
+    clang-tidy \
+    cron \
+    curl \
+    dnsutils \
+    gawk \
+    gdb \
+    git \
+    htop \
+    iproute2 \
+    lcov \
+    libcurl4-openssl-dev \
+    libsecp256k1-dev \
+    libssl-dev \
+    libtool \
+    libxml2-utils \
+    logrotate \
+    net-tools \
+    ninja-build \
+    nload \
+    ocl-icd-opencl-dev \
+    openssh-client \
+    pkg-config \
+    python3-dev \
+    python3-pip \
+    rsync \
+    rsyslog \
+    tar \
+    trickle \
+    unzip \
+    vim \
+    wget \
+    zip \
+    && apt-get remove -y cmake python2  && apt-get autoremove -y
+
+# Cargo
+RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs -o install_script.sh && \
+    sh install_script.sh -y
+ENV PATH="/root/.cargo/bin:${PATH}"
+
+# CMake 3.24
+ARG CMAKE_VERSION=3.24.2
+RUN wget https://github.com/Kitware/CMake/releases/download/v${CMAKE_VERSION}/cmake-${CMAKE_VERSION}-Linux-x86_64.sh \
+    && echo "739d372726cb23129d57a539ce1432453448816e345e1545f6127296926b6754 cmake-${CMAKE_VERSION}-Linux-x86_64.sh" | sha256sum -c \ 
+    && mkdir -p "${HOME}"/.local \
+    && bash ./cmake-${CMAKE_VERSION}-Linux-x86_64.sh --skip-license --prefix="${HOME}"/.local/ \
+    && "${HOME}"/.local/bin/cmake --version \
+    && rm cmake-${CMAKE_VERSION}-Linux-x86_64.sh
+ENV PATH="/root/.local/bin:${PATH}"
+
+# Setup ccache
+RUN ln -s "$(which ccache)" /usr/local/bin/gcc \
+  && ln -s "$(which ccache)" /usr/local/bin/g++ \
+  && ln -s "$(which ccache)" /usr/local/bin/cc \
+  && ln -s "$(which ccache)" /usr/local/bin/c++
+
+# This tag must be equivalent to the hash specified by "builtin-baseline" in vcpkg.json
+ARG VCPKG_COMMIT_OR_TAG=2022.09.27
+ARG VCPKG_ROOT=/vcpkg
 
-# run a copy -L to unfold the symlinkes, and strip all exes
-RUN mkdir -p /scilla/0/bin2/ && cp -L /scilla/0/bin/* /scilla/0/bin2/ && strip /scilla/0/bin2/*
+RUN ccache -p && ccache -z
 
-FROM zilliqa/zilliqa-ccache:${ZILLIQA_CCACHE_VERSION} AS builder
+# If COMMIT_OR_TAG is a branch name or a tag, clone a shallow copy which is
+# faster; if this fails, just clone the full repo and checkout the commit.
+RUN git clone https://github.com/microsoft/vcpkg ${VCPKG_ROOT} \
+  && git -C ${VCPKG_ROOT} checkout ${VCPKG_COMMIT_OR_TAG} \
+  && ${VCPKG_ROOT}/bootstrap-vcpkg.sh
 
-# Manually input tag or commit, can be overwritten by docker build-args
-ARG COMMIT_OR_TAG=v8.4.0
-ARG REPO=https://github.com/Zilliqa/Zilliqa.git
-ARG SOURCE_DIR=/root/zilliqa
-ARG BUILD_DIR=/root/zilliqa/build
-ARG INSTALL_DIR=/usr/local
-ARG VCPKG_ROOT=/vcpkg
+WORKDIR /zilliqa
+
+COPY vcpkg-registry/ vcpkg-registry
+COPY vcpkg-configuration.json .
+COPY vcpkg.json .
+
+RUN ${VCPKG_ROOT}/vcpkg install --triplet=x64-linux-dynamic
```

Comment:
> Compiling the dependencies takes around 20-30 mins

Is there any way to reduce this time? I guess I'll try the `RUN` cache here too, but IMO its less important since our dependency don't change as often as our source code.

> They use a subset of those, but the whole idea is to make sure they use the same base image & dependencies (i.e. environment) as we do.

My idea would be to decouple the dependencies. I don't see why Scilla inherently needs to have the same dependencies as Zilliqa? If anything, as a Scilla developer it would probably annoy me a lot having to built a load of C++ files that I don't use. @jjcnn thoughts?

Reply:
> Is there any way to reduce this time? I guess I'll try the RUN cache here too, but IMO its less important since our dependency don't change as often as our source code.

As I've explained above: the set of dependencies are not stable yet and we've just gone through a phase of updating them pretty frequently. They are quite likely to also change and updated in the future. This would cause a recompilation of this whole thing time and again!

Reply:
> My idea would be to decouple the dependencies. I don't see why Scilla inherently needs to have the same dependencies as Zilliqa? If anything, as a Scilla developer it would probably annoy me a lot having to built a load of C++ files that I don't use. @jjcnn thoughts?

What you're doing is the exact opposite of the whole point of the changes we've introduced in the last couple of months!

Having different dependencies, *especially* when we're using shared libraries all over the place is a pretty bad idea. We've already gone through a phase where Scilla was using different versions of our dependencies (many introduced as old apt packages). On top of one project having a completely different version of something like CryptoUtils it will also have been compiling with lower-level dependencies such as boost, libffi, etc..

I also don't get your argument about "being annoyed as a developer". You have no problem having our docker build for an extra 20-30 mins when our dependencies change (more often) but you find it annoying that for a Scilla developer this compilation happens once (for a reduced subset of those dependencies) and barely ever changes.

Reply:
Just a few comments, without fully understanding what the PR does:
- Building Scilla from scratch takes maybe 20-30 seconds on my machine, so it really isn't prohibitively slow. It depends on the machine of course, but Scilla build time is not something I would consider an issue as of right now.
- Originally the ambition for Scilla was to make it as blockchain-independent as possible, which suggests that the dependencies should be decoupled as much as possible from those of the underlying blockchain. However, as of right now it looks unlikely that Scilla will be adopted by anyone other than us, so that's less of a concern right now. What is a concern right now is manpower shortage, so anything we can do to limit the amount of maintenance required is good by me. If at some point we need to port Scilla to another system, then we can revisit the problem at that time.

Reply:
@jjcnn , @JamesHinshelwood 's argument is mainly about the dependencies (which can take longer to compile, but especially for you guys is really a one-off). 
It might be possible, using static libraries and only relying on the package manager for the underlying OS. This means writing more CMake find scripts and shifting to linking everything using static libraries.
At the moment, given that we use shared libraries all over the place and that we'd like to be able to upgrade the OS / compilers more frequently this isn't easy to do, ensuring the same shared libraries are used by all binaries.